File: src/test/java/PropertyTypeTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import uk.dsx.reactiveconfig.*;

import static org.junit.jupiter.api.Assertions.*;

public class PropertyTypeTest {
    private ReactiveConfig config;
    private ConfigMock source;

    @BeforeEach
    void init(){
        source = new ConfigMock();
        config = new ReactiveConfig.Builder()
                .addSource("configMock", source)
                .build();
    }

    @Test
    void stringNullableTypeTest() throws InterruptedException {
        source.addToMap("nullableStringProperty", null);
        Thread.sleep(10);
        Reloadable nullableStringProperty = config.get("nullableStringProperty", PropertyTypesKt.nullable(PropertyTypesKt.stringType));
        assertNull(nullableStringProperty.get());
    }

    @Test
    void accessingToThePropertyTwiceTest() {
        source.addToMap("property", "something");

        Reloadable property1 = config.get("property", PropertyTypesKt.stringType);
        Reloadable property2 = config.get("property", PropertyTypesKt.stringType);

        assertEquals(property1, property2);
        assertEquals("something", property1.get());
        assertEquals("something", property2.get());
    }

    @Test
    void stringTypeTest() throws InterruptedException {
        source.addToMap("stringProperty", "something1");
        Reloadable<String> stringProperty = config.get("stringProperty", PropertyTypesKt.stringType);

        source.pushChanges("stringProperty", "something2");
        Thread.sleep(10);

        assertEquals("something2", stringProperty.get());
    }

    @Test
    void intTypeTest(){
        source.addToMap("intProperty", 1313);
        Reloadable<Integer> intProperty = config.get("intProperty", PropertyTypesKt.intType);

        assertEquals(1314, intProperty.get() + 1);
    }

    @Test
    void longTypeTest(){
        source.addToMap("longProperty", 1212L);
        Reloadable<Long> longProperty = config.get("longProperty", PropertyTypesKt.longType);

        assertEquals(1212L, longProperty.get());
    }

    @Test
    void floatTypeTest(){
        source.addToMap("floatProperty", 1414F);
        Reloadable<Float>floatProperty = config.get("floatProperty", PropertyTypesKt.floatType);

        assertEquals(1414F, floatProperty.get());
    }

    @Test
    void doubleTypeTest() throws InterruptedException {
        source.addToMap("doubleProperty", 1515.0);
        Reloadable<Double>doubleProperty = config.get("doubleProperty", PropertyTypesKt.doubleType);

        assertEquals(1515.0, doubleProperty.get());
    }

    @Test
    void booleanTypeTest(){
        source.addToMap("booleanProperty", true);
        Reloadable<Boolean> booleanProperty = config.get("booleanProperty", PropertyTypesKt.booleanType);

        assertTrue(booleanProperty.get());
    }
}

File: src/test/kotlin/uk/dsxt/rhea/PropertyTypeTest.kt
package uk.dsxt.rhea

import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.specification.describe
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

object PropertyTypeTest : Spek({
    val source = ConfigMock()

    val config = ReactiveConfig.Builder()
        .addSource("configMock", source)
        .build()

    describe("calling nullable()") {
        source.addToMap("nullableStringProperty", null)
        val nullableStringProperty = config.get("nullableStringProperty", stringType.nullable())

        it("value of property should have been changed to null") {
            assertNull(nullableStringProperty!!.get())
        }
    }

    describe("accessing one property twice") {
        source.addToMap("property", "something")

        val property1 = config.get("property", stringType)
        val property2 = config.get("property", stringType)

        it("the same reloadable should be in both: property1 and property2") {
            assertEquals(property1, property2)
        }
        it("value of property1 should have been changed to 'something'") {
            assertEquals("something", property1!!.get())
        }
        it("value of property2 should also have been changed to 'something'") {
            assertEquals("something", property2!!.get())
        }
    }

    describe("stringType") {
        source.addToMap("stringProperty", "something1")
        val stringProperty = config.get("stringProperty", stringType)

        Thread.sleep(10)
        source.pushChanges("stringProperty", "something2")

            it("value of property should have been changed to 'something2'") {
                assertEquals("something2", stringProperty!!.get())
            }
    }


    describe("intType") {
        source.addToMap("intProperty", 1313)
        val intProperty = config.get("intProperty", intType)

        it("value of property should have been changed to 1313") {
            assertEquals(1314, intProperty!!.get() + 1)
        }
    }

    describe("longType") {
        source.addToMap("longProperty", 1212L)
        val longProperty = config.get("longProperty", longType)

        it("value of property should have been changed to 1212") {
            assertEquals(1212L, longProperty!!.get())
        }
    }

    describe("floatType") {
        source.addToMap("floatProperty", 1414F)
        val floatProperty = config.get("floatProperty", floatType)

        it("value of property should have been changed to 1414") {
            assertEquals(1414F, floatProperty!!.get())
        }
    }

    describe("doubleType") {
        source.addToMap("doubleProperty", 1515.0)
        val doubleProperty = config.get("doubleProperty", doubleType)

        it("value of property should have been changed to 1515") {
            assertEquals(1515.0, doubleProperty!!.get())
        }
    }

    describe("booleanType") {
        source.addToMap("booleanProperty", true)
        val booleanProperty = config.get("booleanProperty", booleanType)

        it("value of property should have been changed to true") {
            assertTrue(booleanProperty!!.get())
        }
    }
})